--- a/nbbuild/netbeans/bin/netbeans	2016-03-05 03:28:28.334856465 +0300
+++ b/nbbuild/netbeans/bin/netbeans	2016-03-05 04:25:09.868809106 +0300
@@ -199,6 +199,9 @@
 	netbeans_default_options="-L-XX:PermSize=32m ${netbeans_default_options}"
 fi

+# Fonts antialiasing
+netbeans_default_options="${netbeans_default_options} -J-Dswing.aatext=TRUE -J-Dawt.useSystemAAFontSettings=on"
+
 launchNbexec() {
     nbexec=`grep -v "^#" "$basedir"/etc/netbeans.clusters | grep -v "^$" | grep platform | while read X; do
         if expr "$X" : "/.*" >/dev/null; then
